File: frameworks/node/h3/UnitTests/cache.test.ts
import assert from "node:assert/strict";
import { test } from "node:test";

import { app, expected } from "./app.ts";

async function serial(url: string, headers: Record<string, string> = {}): Promise<string | null> {
  const response = await app.request(url, { headers });
  return response.headers.get("x-rb-serial");
}

// rb:test cache.small,cache.medium,cache.large
for (const [id, size] of [["cache.small", "small"], ["cache.medium", "medium"], ["cache.large", "large"]] as const) {
  test(`${id}: a second request is the stored answer`, async () => {
    const first = await app.request(`/cache/${size}`);
    const second = await app.request(`/cache/${size}`);

    assert.deepEqual(await second.json(), expected.json(`items.${size}.json`));
    assert.equal(second.headers.get("x-rb-serial"), first.headers.get("x-rb-serial"));
    assert.equal(second.headers.get("x-cache"), "HIT");
    assert.match(second.headers.get("content-type")!, /^application\/json/);
  });
}

// rb:test cache.vary_one
test("cache.vary_one: one vary header keys the store", async () => {
  const alpha = await serial("/cache/vary/one", { "x-rb-tenant": "alpha" });
  const beta = await serial("/cache/vary/one", { "x-rb-tenant": "beta" });

  assert.equal(await serial("/cache/vary/one", { "x-rb-tenant": "alpha" }), alpha);
  assert.notEqual(alpha, beta);
});

// rb:test cache.vary_many
test("cache.vary_many: each of three vary headers keys the store", async () => {
  const webEuAlpha = { "x-rb-channel": "web", "x-rb-region": "eu", "x-rb-tenant": "alpha" };

  const first = await serial("/cache/vary/many", webEuAlpha);

  assert.equal(await serial("/cache/vary/many", webEuAlpha), first);
  assert.notEqual(await serial("/cache/vary/many", { ...webEuAlpha, "x-rb-tenant": "beta" }), first);
  assert.notEqual(await serial("/cache/vary/many", { ...webEuAlpha, "x-rb-region": "us" }), first);
});

test("the answer says what it varies on", async () => {
  const response = await app.request("/cache/vary/many");

  assert.equal(response.headers.get("vary"), "x-rb-channel, x-rb-region, x-rb-tenant");
});

test("a route outside the family is never stored", async () => {
  const first = await serial("/compressed/small");
  const second = await serial("/compressed/small");

  assert.notEqual(second, first);
});

File: frameworks/go/gorilla-mux/UnitTests/cache_test.go
package unittests

import (
	"net/http"
	"testing"
)

// rb:test cache.small,cache.medium,cache.large
func TestASecondRequestIsTheStoredAnswer(t *testing.T) {
	for _, row := range []struct{ id, size string }{
		{"cache.small", "small"},
		{"cache.medium", "medium"},
		{"cache.large", "large"},
	} {
		t.Run(row.id, func(t *testing.T) {
			first := get(t, "/cache/"+row.size)
			second := get(t, "/cache/"+row.size)

			assertOK(t, payload(t, "items."+row.size+".json"), second)
			if serial(t, first) != serial(t, second) {
				t.Fatalf("serial %d then %d, so the handler ran twice", serial(t, first), serial(t, second))
			}
		})
	}
}

// rb:test cache.vary_one
func TestOneVaryHeaderKeysTheStore(t *testing.T) {
	t.Run("cache.vary_one", func(t *testing.T) {
		alpha := serial(t, get(t, "/cache/vary/one", "x-rb-tenant", "alpha"))
		beta := serial(t, get(t, "/cache/vary/one", "x-rb-tenant", "beta"))

		if again := serial(t, get(t, "/cache/vary/one", "x-rb-tenant", "alpha")); again != alpha {
			t.Fatalf("alpha answered %d, then %d", alpha, again)
		}
		if alpha == beta {
			t.Fatal("alpha and beta share one entry")
		}
	})
}

// rb:test cache.vary_many
func TestEachOfThreeVaryHeadersKeysTheStore(t *testing.T) {
	t.Run("cache.vary_many", func(t *testing.T) {
		webEuAlpha := []string{"x-rb-channel", "web", "x-rb-region", "eu", "x-rb-tenant", "alpha"}
		webEuBeta := []string{"x-rb-channel", "web", "x-rb-region", "eu", "x-rb-tenant", "beta"}

		response := get(t, "/cache/vary/many", webEuAlpha...)
		first := serial(t, response)

		assertOK(t, payload(t, "items.small.json"), response)
		if again := serial(t, get(t, "/cache/vary/many", webEuAlpha...)); again != first {
			t.Fatalf("web, eu, alpha answered %d, then %d", first, again)
		}
		if serial(t, get(t, "/cache/vary/many", webEuBeta...)) == first {
			t.Fatal("alpha and beta share one entry")
		}
	})
}

func TestTheAnswerSaysWhatItVariesOn(t *testing.T) {
	response := get(t, "/cache/vary/many")

	assertStatus(t, response, http.StatusOK)
	if vary := response.Header.Get("Vary"); vary != "x-rb-channel, x-rb-region, x-rb-tenant" {
		t.Fatalf("vary %q", vary)
	}
}
